elastic/apm-server
# Introduce `apm-server.auth.anonymous` config

## Motivation/summary

Generalise the ability for agents to send events unauthenticated (anonymous) but rate-limited. Until now this has been a RUM-only feature, but we now find ourselves needing it also for the iOS agent.

Anonymous auth is disabled by default, but is automatically enabled when RUM is enabled as long as `apm-server.auth.anonymous` hasn't been explicitly configured. The existing RUM config for allowed services and rate limiting are deprecated and replaced with equivalent config under `apm-server.auth.anonymous.*`.

Instead of restricting anonymous auth to requests going by endpoint (i.e. RUM intake and agent config), we now restrict based on the provided agent and service names. There was previously nothing stopping clients from spoofing RUM agents, e.g. sending events to the RUM intake with a non-RUM agent name, so this is not any less secure.

## Checklist


## How to test these changes

1. Ensure existing config works as before: RUM should be allowed to send data without any auth token when secret_token or api_key auth is enabled; ensure no other agents can send data without an auth token.
2. Define `apm-server.auth.anonymous.allow_agent: [iOS/swift]`, check that the iOS/swift agent can send data without an auth token.
3. Define `apm-server.auth.anonymous.allow_service: [opbeans-rum]`, check that opbeans-rum can send data. Change it to something else and check that opbeans-rum cannot send data.

## Related issues

Closes https://github.com/elastic/apm-server/issues/5347
